rai/node/rpc.cpp

```diff
@@ -889,6 +889,127 @@ void rai::rpc_handler::accounts_pending ()
 	response (response_l);
 }
 
+void rai::rpc_handler::sign_message ()
+{
+	rai::uint256_union account (0);
+	boost::optional<std::string> account_text (request.get_optional<std::string> ("account"));
+	if (account_text.is_initialized ())
+	{
+		auto error_account (account.decode_account (account_text.get ()));
+		if (error_account)
+		{
+			error_response (response, "Bad account number");
+		}
+	}
+	rai::uint256_union wallet (0);
+	boost::optional<std::string> wallet_text (request.get_optional<std::string> ("wallet"));
+	if (wallet_text.is_initialized ())
+	{
+		auto error (wallet.decode_hex (wallet_text.get ()));
+		if (error)
+		{
+			error_response (response, "Bad wallet number");
+		}
+	}
+	rai::raw_key prv;
+	prv.data.clear ();
+	if (wallet != 0 && account != 0)
+	{
+		auto existing (node.wallets.items.find (wallet));
+		if (existing != node.wallets.items.end ())
+		{
+			rai::transaction transaction (node.store.environment, nullptr, false);
+			auto unlock_check (existing->second->store.valid_password (transaction));
+			if (unlock_check)
+			{
+				auto account_check (existing->second->store.find (transaction, account));
+				if (account_check != existing->second->store.end ())
+				{
+					existing->second->store.fetch (transaction, account, prv);
+				}
+				else
+				{
+					error_response (response, "Account not found in wallet");
+				}
+			}
+			else
+			{
+				error_response (response, "Wallet is locked");
+			}
+		}
+		else
+		{
+			error_response (response, "Wallet not found");
+		}
+	}
+	boost::optional<std::string> key_text (request.get_optional<std::string> ("key"));
+	if (key_text.is_initialized ())
+	{
+		auto error_key (prv.data.decode_hex (key_text.get ()));
+		if (error_key)
+		{
+			error_response (response, "Bad private key");
+		}
+	}
+	std::string message (request.get<std::string> ("message"));
+	if (prv.data != 0)
+	{
+		message.insert (0, "Nano Signed Message:\n");
+		rai::uint256_union message_hash;
+		blake2b_state hash;
+		blake2b_init (&hash, message_hash.bytes.size ());
+		blake2b_update (&hash, message.data (), message.length ());
+		blake2b_final (&hash, message_hash.bytes.data (), message_hash.bytes.size ());
+		rai::uint256_union pub;
+		ed25519_publickey (prv.data.bytes.data (), pub.bytes.data ());
+		rai::uint512_union signature = rai::sign_message (prv, pub, message_hash);
```

Comment:
I think with ed25519 it's better to sign the message itself instead of the hash. Ed25519 will already internally hash the message. We don't do this throughout the rest of the codebase, but signing isn't inherently related to Nano so I'd prefer to stick with convention here.

Also applies to `verify_message` of course.